WalletWasabi/Crypto/Api/WabiSabiClient.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using NBitcoin;
+using NBitcoin.Secp256k1;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Crypto.ZeroKnowledge.LinearRelation;
+using WalletWasabi.Crypto.ZeroKnowledge.NonInteractive;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.Api
+{
+	public class WabiSabiClient : ICredentialRequestBuilder
+	{
+		private CoordinatorParameters CoordinatorParameters { get; }
+		private readonly WasabiRandom Random;
+		private readonly List<Money> CredentialsToRequest = new List<Money>(); 
+		private readonly List<(Money Amount, CredentialIssuanceRequest Credential, Scalar r, Knowledge Knowledge)> CredentialsRequested = new List<(Money, CredentialIssuanceRequest, Scalar, Knowledge)>(); 
+		private readonly List<(Credential Credential, Scalar z, Knowledge Knowledge)> CredentialsToPresent = new List<(Credential, Scalar, Knowledge)>();
+
+		public WabiSabiClient(CoordinatorParameters coordinatorParameters, WasabiRandom rnd)
+		{
+			Random = Guard.NotNull(nameof(rnd), rnd);
+			CoordinatorParameters = Guard.NotNull(nameof(coordinatorParameters), coordinatorParameters);
+		}
+
+		private Money Balance => CredentialsToRequest.Sum() - Credentials.Sum(x => x.Amount.ToMoney());
+
+		private bool IsNullRequest => Balance == Money.Zero && !CredentialsToPresent.Any();
+
+		public List<Credential> Credentials { get; } = new List<Credential>();
+
+		public void HandleResponse(RegistrationResponse registrationResponse)
+		{
+			Guard.NotNull(nameof(registrationResponse), registrationResponse);
+
+			var requestedCredentials = CredentialsRequested;
+			var issuedCredentials = registrationResponse.IssuedCredentials;
+
+			if (issuedCredentials.Count() != requestedCredentials.Count())
+			{
+				throw new WabiSabiException(
+					WabiSabiErrorCode.IssuedCredentialNumberMismatch, 
+					$"{issuedCredentials.Count()} issued but {requestedCredentials.Count()} were requested.");
+			}
+			var credentials = Enumerable
+				.Zip(requestedCredentials, issuedCredentials)
+				.Select(x => new { Requested = x.First, Issued = x.Second })
+				.ToArray();
+
+			var statements = credentials
+				.Select(x => ProofSystem.IssuerParameters(CoordinatorParameters, x.Issued, x.Requested.Credential.Ma));
+
+			var areCorrectlyIssued = Verifier.Verify(new Transcript(new byte[0]), statements, registrationResponse.Proofs);
```

Comment:
here too i think the transcript should be [reused](https://merlin.cool/use/duality.html) for both requests and responses